feat: pull completed program stage event flow runs

JohanGHole opened the pull request:
Having some troubles with triggering flows from the connector, and I have to look into if the RapidPro rate limiting will be an issue. In the meantime I'll finish up the RapidPro => DHIS2 data flow (fetching and transforming flow runs to dhis2 events). 

This PR:
* refactor: `jms:queue:dhis2` => `jms:queue:dhis2AggregateReports`
* refactor: rename `PullReportsRouteBuilder` to `PullRapidProFlowsRouteBuilder`
* feat: include pulling program stage event flows in `PullRapidProFlowsRouteBuilder`
* feat: add `SetFlowUuidPropertyProcessor` to concatenate aggregate and event flow uuids
* test: added and extended tests in `PullRapidProFlowsRouteBuilderFunctionalTestCase`
* feat: simple datasonnet transformation for completed event flow runs
* feat: created queue: `jms:queue:dhis2ProgramStageEvents`

Next PR:
* add `DeliverEventsRouteBuilder` with error handling (dlq support) and event transformation. 
* Update README

JohanGHole committed:
feat: pull completed program stage event flow runs

claudemamo commented on src/test/java/org/hisp/dhis/integration/rapidpro/route/PullRapidProFlowsRouteBuilderFunctionalTestCase.java:
```diff
@@ -129,39 +147,153 @@ private DataValue getDataValue( String dataElementId, DataValueSet dataValueSet
+        camelContext.start();
+
+        producerTemplate.sendBody( "direct:pull", ExchangePattern.InOnly, null );
+        Thread.sleep( 15000 );
```
Not a fan of `Thread.sleep(...)`. What about `spyEndpoint.assertIsNotSatisfied(...)`?

JohanGHole replied:
Sure. Will the replace `Thread.sleep(...)` with `mockEndpoint.await(...)` and / or `mockEndpoint.assertIsSatisified(...)`

claudemamo commented on src/test/java/org/hisp/dhis/integration/rapidpro/route/PullRapidProFlowsRouteBuilderFunctionalTestCase.java:
```diff
@@ -129,39 +147,153 @@ private DataValue getDataValue( String dataElementId, DataValueSet dataValueSet
+        camelContext.start();
+        runFlowAndWaitUntilCompleted( programStageEventFlowUuid, eventId );
+        producerTemplate.sendBody( "direct:pull", ExchangePattern.InOnly, null );
+        Thread.sleep( 15000 );
```
See my previous comment. What about `assertIsSatisfied(spyEndpoint)`?

claudemamo commented on src/test/java/org/hisp/dhis/integration/rapidpro/route/PullRapidProFlowsRouteBuilderFunctionalTestCase.java:
```diff
@@ -129,39 +147,153 @@ private DataValue getDataValue( String dataElementId, DataValueSet dataValueSet
+            Thread.sleep( 1000 );
+        }
+        producerTemplate.sendBody( "direct:pull", ExchangePattern.InOnly, null );
+        Thread.sleep( 10000 );
```
See my previous comment.

claudemamo commented on src/test/java/org/hisp/dhis/integration/rapidpro/route/PullRapidProFlowsRouteBuilderFunctionalTestCase.java:
```diff
@@ -129,39 +147,153 @@ private DataValue getDataValue( String dataElementId, DataValueSet dataValueSet
+        syncTrackedEntityContact( "whatsapp:12345678" );
 
         producerTemplate.sendBody( "direct:pull", ExchangePattern.InOnly, null );
         Thread.sleep( 15000 );
```
See my previous comment.

claudemamo commented on src/main/java/org/hisp/dhis/integration/rapidpro/processor/SetFlowUuidPropertyProcessor.java:
```diff
@@ -0,0 +1,69 @@
+{
+
+    @Autowired
+    ProgramStageToFlowMap programStageToFlowMap;
```
Should this be protected or private?

claudemamo commented on src/main/java/org/hisp/dhis/integration/rapidpro/processor/SetFlowUuidPropertyProcessor.java:
```diff
@@ -0,0 +1,69 @@
+        Exception
+    {
+        String aggregateReportFlowUuids = exchange.getContext()
+            .resolvePropertyPlaceholders( "{{rapidpro.flow.uuids:}}" );
```
Isn't this set from the property file? You should be able to inject it with @<!-- -->Value.

claudemamo commented on src/main/java/org/hisp/dhis/integration/rapidpro/processor/SetFlowUuidPropertyProcessor.java:
```diff
@@ -0,0 +1,69 @@
+    {
+        String aggregateReportFlowUuids = exchange.getContext()
+            .resolvePropertyPlaceholders( "{{rapidpro.flow.uuids:}}" );
+        String programStageEventFlowUuids = programStageToFlowMap.getFlowUuids();
+        if ( aggregateReportFlowUuids.isEmpty() && !programStageEventFlowUuids.isEmpty() )
+        {
+            exchange.setProperty( "flowUuids", programStageEventFlowUuids );
+        }
+        else if ( !aggregateReportFlowUuids.isEmpty() && programStageEventFlowUuids.isEmpty() )
+        {
+            exchange.setProperty( "flowUuids", aggregateReportFlowUuids );
+        }
+        else
+        {
+            exchange.setProperty( "flowUuids",
+                String.join( ",", aggregateReportFlowUuids, programStageEventFlowUuids ) );
+        }
+    }
```
I might be missing something but can this `if` statement be replaced with just:

```java
 exchange.setProperty( "flowUuids",
                String.join( ",", aggregateReportFlowUuids, programStageEventFlowUuids ) );
```

claudemamo replied:
And if it's going to be a one-liner, might be worth considering pulling this up to the route.

JohanGHole replied:
Solved it with a minor change in the `programStageToFlowMap `implementation. Can pull it up to the route.

claudemamo commented on src/main/java/org/hisp/dhis/integration/rapidpro/route/PullRapidProFlowsRouteBuilder.java:
```diff
@@ -77,20 +82,40 @@ protected void doConfigure()
+        from( "direct:queueProgramStageEvent" )
+            .routeId( "Queue Program Stage Event" )
+            .setHeader( "eventId", simple( "${body[values][event_id][value]}" ) )
+            .transform( datasonnet( "resource:classpath:eventFlowRun.ds", String.class, "application/x-java-object",
```
Use DataSonnet just for creating an entry seems like an overkill to me. Would `Map.of(...)` achieve the same thing?

JohanGHole replied:
Replaced `eventFlowRun.ds` with `webhook.ds` to make the payload consistent with the aggregate report case. Might want to change to a JMS selector eventually, especially if we are going to add more features.

cjmamo reviewed (commented):
Added a no. of comments. In particular, we should really try to stay away from Thread.sleep(...), especially in production code. Regarding the queues, not sure if it makes sense to use JMS selectors instead of creating additional queues. Something you might want to look into.

JohanGHole committed:
fix: removed Thread.sleep(...) from tests

Replaced with mockEndpoint.await(...) and mockEndpoint.assertIsSatisified(...)

JohanGHole committed:
refactor: replace SetFlowUuidPropertyProcessor with inline lambda processor

JohanGHole committed:
chore: bump sdk to latest version

JohanGHole committed:
fix: increase timeout

JohanGHole committed:
fix: CI build issue

JohanGHole committed:
refactor: replaced eventFlowRun.ds with webhook.ds

JohanGHole commented:
* Removed all `Thread.Sleep(...) ` calls from the tests, except where it is needed (i.e. waiting for completed flow runs)
* Increased timeout values to fix CI build issues. Might still be too conservative, so will keep an eye on this
* Replaced the `SetFlowUuidPropertyProcessor` with an inline lambda processor
* Minor fixes

Also:
I've taken note of your suggestion to utilize JMS selectors instead of setting up new queues. I will investigate this a bit further while working on the implementation of the deliver event route builder. If this change leads to a more streamlined and clean pull flow route builder, I will include it in the next PR.

claudemamo commented on src/main/java/org/hisp/dhis/integration/rapidpro/ProgramStageToFlowMap.java:
```diff
@@ -83,6 +83,10 @@ public String getFlowUuids( Map<String, Object> body )
 
     public String getFlowUuids()
     {
+        if ( this.map.values().isEmpty() )
+        {
+            return "";
+        }
         return String.join( ",", this.map.values() );
```
This can be reduce to a ternary operator.

claudemamo commented on src/test/java/org/hisp/dhis/integration/rapidpro/Environment.java:
```diff
@@ -158,11 +158,11 @@ private static void setUpDhis2()
+            .withKeepAliveDuration( 500000L, TimeUnit.MILLISECONDS )
             .withCallTimeout( 0L, TimeUnit.MILLISECONDS )
-            .withReadTimeout( 20000L, TimeUnit.MILLISECONDS )
-            .withWriteTimeout( 20000L, TimeUnit.MILLISECONDS )
-            .withConnectTimeout( 10000L, TimeUnit.MILLISECONDS )
+            .withReadTimeout( 30000L, TimeUnit.MILLISECONDS )
+            .withWriteTimeout( 30000L, TimeUnit.MILLISECONDS )
+            .withConnectTimeout( 20000L, TimeUnit.MILLISECONDS )
```
I'm wondering whether we need to set all of these. Which is causing the timeout?

JohanGHole replied:
My guess is the `keepAliveDuration` in combination with some of the `spyEndpoint.await(...) ` method calls being too conservative. At least that was the case in `WebhookFunctionalTestCase `which caused the CI build issues. Let me have another look.

claudemamo commented on src/main/java/org/hisp/dhis/integration/rapidpro/route/PullRapidProFlowsRouteBuilder.java:
```diff
@@ -61,8 +65,13 @@ protected void doConfigure()
+            .process( exchange -> {
+                exchange.setProperty( "flowUuids",
+                    String.join( ",", programStageToFlowMap.getFlowUuids(),
+                        exchange.getContext().resolvePropertyPlaceholders( "{{rapidpro.flow.uuids:}}" ) ) );
```
`rapidpro.flow.uuids` should be injected (see the `Value` annotation).

JohanGHole replied:
I've previously tried injecting it using `@Value`. It works fine in production, but it causes issues in the test environment, since the bean is initialized before the `rapidpro.flow.uuids` property values is set in the tests. Do you know if there is a simple workaround for this? Looked into creating a custom context initializer, but I feel this might be an overkill for such a seemingly trivial task.

claudemamo replied:
IIRC, I used `System.setProperty(...)` before starting the Camel context in the test method in order for  `@Value` to work.

JohanGHole replied:
I inject it as follows:
``` 
@Value( "${rapidpro.flow.uuids}" )
private String aggregateReportFlowUuids;
```
And set the property in the tests as follows, making sure that all of the properties are set before starting the Camel context:
```
        System.setProperty( "sync.rapidpro.contacts", "true" );
        System.setProperty( "rapidpro.flow.uuids", aggregateReportFlowUuid );
        AdviceWith.adviceWith( camelContext, "Transmit Report", r -> r.weaveAddLast().to( "mock:spy" ) );
        MockEndpoint spyEndpoint = camelContext.getEndpoint( "mock:spy", MockEndpoint.class );
        spyEndpoint.setExpectedCount( 0 );
        camelContext.start();
        ...
```
Still no luck. `Rapidpro.flow.uuids` is still empty. Had the same issue in the first iteration of the connector when trying to inject the program stage codes. Ended up having to use `setProperty("programStageId", simple("{{program.stage.id}}")` in the route.

JohanGHole replied:
Think I found something here:
https://stackoverflow.com/questions/17353327/populating-spring-value-during-unit-test
Will let you know how it goes.

cjmamo reviewed (commented):
Added some final comments. Other than that you're good to merge.

JohanGHole committed:
refactor: resolve PR comments

- Replaced `resolvePropertyPlaceholder` with `@Value` injection.
- Reverting unneeded timeout changes.
- Added ternary operator in the `flowUuid` concatenator to avoid exception when both `aggregateFlowUuids` and `eventFlowUuids` are empty.

JohanGHole commented:
- Replaced `resolvePropertyPlaceholder` with `@Value` injection.
- Reverting unneeded timeout changes.
- Added ternary operator in the `flowUuid` concatenator processor to avoid exception thrown when both `aggregateFlowUuids` and `eventFlowUuids` are empty.

JohanGHole committed:
fix: Removed redundant timeout settings in client configuration

JohanGHole committed:
refactor: removed withMaxIdleConnections

JohanGHole committed:
Merge branch 'main' into pull-flows-route-builder

JohanGHole committed:
refactor: timeouts

JohanGHole committed:
revert: revert latest timeout changes

JohanGHole committed:
chore: increase connection timeout

JohanGHole committed:
chore: updated timeouts

JohanGHole committed:
chore: increase call timeout

JohanGHole committed:
chore: bump actions/checkout from 3 to 4

JohanGHole committed:
chore: increase timeouts

JohanGHole committed:
chore: update connection timeout

JohanGHole committed:
Merge branch 'main' into pull-flows-route-builder

JohanGHole committed:
chore: update timeouts
